src/odfesqlodbc/es_communication.cpp

```diff
@@ -438,9 +473,65 @@ int ESCommunication::ExecDirect(const char* query, const char* fetch_size_) {
         return -1;
     }
     m_result_queue.push(std::unique_ptr< ESResult >(result));
+    if (!result->cursor.empty()) {
+        // If response has cursor, this thread will retrives more results pages asynchronously.
+        auto send_cursor_queries = std::async(std::launch::async, [&]() {
+            SendCursorQueries(result->cursor.c_str());
+        });
+    }
     return 0;
 }
 
+void ESCommunication::SendCursorQueries(const char* _cursor) {
+    if (_cursor == NULL)
+        return;
+    try {
+        std::string cursor(_cursor);
+        while (!cursor.empty() && (m_result_queue.size() < m_result_queue_capacity)) {
```

Comment:
This seems like it will simply stop when the queue reaches capacity. A better approach would be to not expose the capacity, but to have the queue be created with a capacity. When you try to push onto the queue when it's at capacity, you block.

Reply:
Will resolve this in separate PR.
Issue: #105